kore/src/Kore/Step/SMT/Evaluator.hs

```diff
@@ -173,30 +180,114 @@ goTranslatePredicate
     => MonadLog m
     => SExpr  -- ^ type name
-    -> p  -- ^ uninterpreted pattern
-    -> Translator m p SExpr
-translateUninterpreted t pat =
-    lookupPattern <|> freeVariable
+    -> TranslateItem variable  -- ^ uninterpreted pattern
+    -> Translator m variable SExpr
+translateTerm smtType (QuantifiedVariable var) = do
+    n <- Counter.increment
+    let varName = "<" <> Text.pack (show n) <> ">"
+        smtVar = SimpleSMT.const varName
+    State.modify'
+        (Lens.over (field @"quantifiedVars") $ Map.insert var
+            SmtEncoding
+            { smtName = varName
+            , smtType
+            , boundVars = []
+            }
+        )
+    return smtVar
+translateTerm t (UninterpretedTerm (TermLike.ElemVar_ var)) =
+    lookupVariable var <|> declareVariable t var
+translateTerm t (UninterpretedTerm pat) = do
+    VarContext { quantifiedVars, terms } <- State.get
+    let freeVars = getFreeVariables $ TermLike.freeVariables pat
+        boundVarsMap =
+            Map.filterWithKey
+                (\k _ -> ElemVar k `Set.member` freeVars)
+                quantifiedVars
+        boundVars = Map.keys boundVarsMap
+        boundPat = TermLike.mkExistsN boundVars pat
+    lookupUninterpreted boundPat quantifiedVars terms
+        <|> declareUninterpreted boundPat boundVars boundVarsMap
   where
-    lookupPattern = do
-        result <- State.gets $ Map.lookup pat
-        maybe empty (return . fst) result
-    freeVariable = do
+    lookupUninterpreted boundPat quantifiedVars terms =
+        maybe empty (translateSmtEncoding quantifiedVars)
+        $ Map.lookup boundPat terms
+    declareUninterpreted boundPat boundVars boundVarsMap
+      = do
         n <- Counter.increment
-        var <- SMT.declare ("<" <> Text.pack (show n) <> ">") t
-        State.modify' (Map.insert pat (var, t))
-        logVariableAssignment n
-        return var
-    logVariableAssignment n =
-        logDebug
-        . Text.pack . show
-        . Pretty.nest 4 . Pretty.sep
-        $ [Pretty.pretty n, "|->", unparse pat]
+        logVariableAssignment n boundPat
+        let smtName = "<" <> Text.pack (show n) <> ">"
+            cache = SmtEncoding { smtName, smtType = t, boundVars }
+        _ <- SMT.declareFun
+            SMT.FunctionDeclaration
+            { name = smtName
+            , inputSorts =
+                smtType <$> mapMaybe (`Map.lookup` boundVarsMap) boundVars
+            , resultSort = t
+            }
+        State.modify'
+            ( Lens.over (field @"terms") $ Map.insert boundPat cache )
+        translateSmtEncoding boundVarsMap cache
+
+lookupVariable
+    :: InternalVariable variable
+    => SMT.MonadSMT m
+    => TermLike.ElementVariable variable
+    -> Translator m variable SExpr
+lookupVariable var = do
+    VarContext { freeVars, quantifiedVars } <- State.get
+    maybe empty return $
+        (SMT.Atom . smtName <$> Map.lookup var quantifiedVars)
+        <|>
+        Map.lookup var freeVars
```

Comment:
Even though this function is short, I had to read it through several times. I found it difficult to follow because the handling of free and quantified variables is intertwined.

```suggestion
lookupVariable var =
    lookupQuantifiedVariable <|> lookupFreeVariable
  where
    lookupQuantifiedVariable = do
        VarContext { quantifiedVars } <- State.get
        hoistMaybe $ SMT.Atom . smtName <$> Map.lookup var quantifiedVars
    lookupFreeVaraible = do
        VarContext { freeVars } <- State.get
        hoistMaybe $ Map.lookup var freeVars
```